external-crates/move/crates/move-vm-runtime/src/execution/dispatch_tables.rs

```diff
@@ -477,78 +307,182 @@ impl VMDispatchTables {
     }
 
     // -------------------------------------------
-    // Type Translation Helpers
+    // Type Depth Computations
     // -------------------------------------------
-    // Note that these are all "lazy": they only
-    // fill out datatype information fields as
-    // they are requested, not before.
+    // These functions compute the depth of the specified type, plus cache the depth formula of any
+    // uncached subtypes they find along the way.
+
+    fn cached_type_depth(&self, datatype: &VirtualTableKey) -> Option<&DepthFormula> {
+        let package = self.type_depths.get(&datatype.package_key)?;
+        package.get(&datatype.inner_pkg_key)
+    }
+
+    pub fn calculate_depth_of_type(
+        &mut self,
+        datatype_name: &VirtualTableKey,
+    ) -> PartialVMResult<DepthFormula> {
+        let mut depth_cache = BTreeMap::new();
+        let depth_formula =
+            self.calculate_depth_of_datatype_and_cache(datatype_name, &mut depth_cache)?;
+        for (datatype_key, depth) in depth_cache {
+            match self
+                .type_depths
+                .entry(datatype_key.package_key)
+                .or_default()
+                .entry(datatype_key.inner_pkg_key)
+            {
+                std::collections::btree_map::Entry::Vacant(vacant_entry) => {
+                    vacant_entry.insert(depth);
+                }
+                std::collections::btree_map::Entry::Occupied(occupied_entry) => {
+                    // This could happen if we race for filling in the depth of the datatype which
+                    // is fine as only one will win. However, since a dispatch table should always
+                    // be in a single tear-off VM, this seems unlikely to happen.
```

Comment:
This seems completely impossible to me unless I'm misunderstanding something, so I think I would be on the "assert the insert is `None`" side of things here probably.

Reply:
I also agree this should be impossible, so have swapped it to do that.